packages/iov-bns/src/types.ts

```diff
@@ -23,6 +23,89 @@ export interface CashConfiguration {
   readonly minimalFee: Amount;
 }
 
+// Governance
+
+export interface Elector {
+  readonly address: Address;
+  /** The voting weight of this elector. Max value is 65535 (2^16-1). */
+  readonly weight: number;
+}
+
+export interface Electorate {
+  readonly id: number;
+  readonly version: number;
+  readonly admin: Address;
+  readonly title: string;
+  /** An unordered map from address to elector */
+  readonly electors: { readonly [index: string]: Elector };
+  /** Sum of all electors' weights */
+  readonly totalWeight: number;
+}
+
+export interface Fraction {
+  readonly numerator: number;
+  readonly denominator: number;
+}
+
+export interface ElectionRule {
+  readonly id: Uint8Array;
+  readonly version: number;
+  readonly admin: Address;
+  /** The electorate used by this (ID without version, as changing electorate changes the rule) */
+  readonly electorateId: number;
+  readonly title: string;
+  /** Voting period in seconds */
+  readonly votingPeriod: number;
+  readonly threshold: Fraction;
+  readonly quorum: Fraction | null;
+}
+
+export interface VersionedId {
+  readonly id: Uint8Array;
+  readonly version: number;
+}
+
+export enum ProposalExecutorResult {
+  NotRun,
+  Succeeded,
+  Failed,
+}
+
+export enum ProposalResult {
+  Undefined,
+  Accepted,
+  Rejected,
+}
+
+export enum ProposalStatus {
+  Submitted,
+  Closed,
+  Withdrawn,
+}
+
+/** Union type for possible options */
+export type ProposalOption = string;
+
+export interface Proposal {
+  readonly title: string;
+  /** The transaction to be executed when the proposal is accepted */
+  readonly option: ProposalOption;
+  readonly description: string;
+  readonly electionRule: VersionedId;
+  readonly electorate: VersionedId;
+  /** Time when the voting on this proposal starts (Unix timestamp) */
+  readonly votingStartTime: number;
+  /** Time when the voting on this proposal starts (Unix timestamp) */
+  readonly votingEndTime: number;
+  /** Time of the block where the proposal was added to the chain (Unix timestamp) */
+  readonly submissionTime: number;
```

Comment:
Do we need to use this type to e.g. preview proposals before submission? If so we might want to make this optional.

Reply:
The blockchain is the only place where `Proposal` is created. An instruction to create a proposal is the `CreateProposalTx`, which should be the only source of data for any kind of preview.